Optimise logic to add BDs to a RaidGroup in a CSPC to maximise successful pool provisions

vharsh opened the pull request:
As of now, the minimum number of Blockdevices get added into a single RAIDgroup.

TODO: ~Consider storage sizes for creating a pool~.

# Implementation to minimize disk losses

- For stripe, there can be only one Raid-Group(in a CSPI) where all the disks exist and they can be of any sizes
- For Mirror, RaidZ, RaidZ2 the BD sizes must be the same in a single `CSPI.RaidGroup`( same size exact to the number of bytes, one byte off and the CSPC will not provision because of a ZFS error)

## How the BD selection works for non-stripe pools?

### Tl; dr summary

1. Filter BDs by minimumSize, Unclaimed+Active states
2. Sort them by increasing order of size
3. Assume, minimum device-level replication for mirror, raidZ, Raidz2(i.e. they'll always have `2, 3 and 6` devices per-RAIDgroup)
4. Select disks from the list^ of disks available in the node

### Step-4 in detail

1. Have a sorted slice of BDs
2. Convert the slice of BDs into a singly-linked-list
3. Run a two-pointer algorithm similar to Sliding Window, the first pointer(i.e. `curr`) will point to the start of the window and the second pointer(i.e. `ahead`) will point to the end of the window, including both the boundaries, they'd always have `1, 2, 3, 6` boxes in their bounds.
4. If the sizes of the BD pointed by the `curr` BD and the `ahead` BD are same, we can confirm that the n elements between them are of the same size, where `n ∀ n ∈ {0, 1, 2..}`
NOTE: If the `curr` & `ahead` point to the same BD node, the single BD is selected

5. We can get rid of those BD nodes by keeping track of the `prev` pointer to the BD pointed to by the `curr` BD
  - the edge cases around selection of the first BD is handled by prefixing a dummy useless node in the `Select(...)` method which is never shared with the caller in any of the edge cases
6. If a chosen minimum number of same sized BDs aren't found between the two pointers, we can confirm that a specific node doesn't have enough BDs for that RaidGroup in a particular pool-instance and promptly throw an error

## Future possible enhancements

- Add a feature flag for omitting HDDs, sparse-files, selecting only HDDs, etc by just modifying `filterCStorCompatibleBDs` function
- Add some automatically fuzz-ed unit tests to further harden the logic


Co-Authored-by: Sai Chaithanya (@mittachaitu) <sai.chaithanya@mayadata.io>
Signed-off-by: Harsh Vardhan (@vharsh) <harsh.vardhan@mayadata.io>

Commit:
Refactor logic to add BDs to a RaidGroup in a CSPC

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Add unit-tests for RaidZ & RaidZ2 pool-type

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Add negative testcases for makePools

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Remove redundant checks post testify checks

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

vharsh commented on pkg/generate/cspc.go:
```diff
@@ -181,10 +182,14 @@ func makePools(poolType string, nDevices int, bd map[string][]v1alpha1.BlockDevi
+	case string(cstorv1.PoolMirrored), string(cstorv1.PoolRaidz), string(cstorv1.PoolRaidz2):
+		// 1. Check for devices per node count against the fixed minimum
+		min := minCount()[poolType]
+		if nDevices < min {
```
Currently there's a check for minimum number(n=1) of BDs in a RaidGroup, more inputs will be needed to pass the webhook validation of `2^n + 1` & `2^n + 2` BDs in a RaidGroup.

vharsh commented on pkg/generate/cspc.go:
```diff
@@ -193,32 +198,36 @@ func makePools(poolType string, nDevices int, bd map[string][]v1alpha1.BlockDevi
+				var raids []cstorv1.CStorPoolInstanceBlockDevice
+				for j := 0; j < min; j++ {
+					raids = append(raids, cstorv1.CStorPoolInstanceBlockDevice{BlockDeviceName: bds[index].Name})
+					index++
```
I can just copy-paste the validation logic for `2^n +1` or `2^n +2` BDs per RaidGroup.

mittachaitu replied:
Yes, You can do that

vharsh replied:
I need to think how I'll implement stacking similarly sized BDS in one RaidGroup now. Sorting might be a good trivial solution for the same.

mittachaitu replied:
Yes, we can do sort of BlockDevices based on their size in descending order and start picking by index but this will not fit in all cases let's say userA has the following bds of size {100, 50, 50, 20} and if he requested to create mirror pool then we need to pick {50, 50} size disk then only user can get pool size without much wastage. 

**How about doing it in the following way**?
- Since we have a list of eligible blockdevices on a host, we will form blodeviceList based on their size(Keep note of sizes in descending order).
- Now let's pick the bucket which satisfies `min` no.of blockdevices and add them to the pool. Once after adding them to the pool remove them from a bucket.

Corner case we need to handle:
- If we are unable to fill up the minimum no.of blockdevices per raid group then add unpicked blockdevices to the raidGroup based on the size in decresing order.

vharsh replied:
Okay, I'll have to approximate and create storage size buckets to hold similar sizes together and see if the applied `nDevices` can be given in a CSPC.

mittachaitu replied:
Yes

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -181,10 +182,14 @@ func makePools(poolType string, nDevices int, bd map[string][]v1alpha1.BlockDevi
-			return nil, fmt.Errorf("mirrored pool requires multiples of two block device")
+	case string(cstorv1.PoolMirrored), string(cstorv1.PoolRaidz), string(cstorv1.PoolRaidz2):
+		// 1. Check for devices per node count against the fixed minimum
+		min := minCount()[poolType]
+		if nDevices < min {
+			return nil, fmt.Errorf("%s pool requires a minimum of %d block device per node",
+				poolType, min)
 		}
```
Since `nDevices` is user input instead of validating here, better to validate  at initial phase itself [here](https://github.com/openebs/openebsctl/blob/92f3001d93c34dabea0fcfbce5ed19d80dc8266e/pkg/generate/cspc.go#L46)

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -193,32 +198,36 @@ func makePools(poolType string, nDevices int, bd map[string][]v1alpha1.BlockDevi
 			index := 0
-			for d := 0; d < nDevices/2; d++ {
-				raids := []cstorv1.CStorPoolInstanceBlockDevice{{BlockDeviceName: bds[index].Name},
-					{BlockDeviceName: bds[index+1].Name}}
```
Q: Do we need to log some info when `nDevices%min != 0`? Informing only `x` devices has picked for pool creation on so and so host?

vharsh replied:
Right, this makes sense, I can just add this check assuming minimum number of disks per RAID-group. As, adding more and more BDs than the minimum will just increase the replication and not the pool-size.

Commit:
Add capacity flag for CSPC generation

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Make RaidZ2 accept minimum 6 disks per node

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
MidPoint-checkin: LinkedList+Sliding window logic

NOTE: Two of unit-tests are failing, which I'll be debugging and fixing
soon-ish.

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Fix wrongly generated BD-list from bdGen()

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Implement the LinkedList approach for iteratively selecting BDs

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Fixed failing unit test

- Same BD was re-used in subsequent RAID-group
- Reason: Assignment to the method receivers propagates
  only to callees but not to callers.

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Remove unused variable `index`

Now BDs come from `Select()` from inside a LinkedList and not an array.

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

mittachaitu commented on cmd/generate/generate.go:
```diff
@@ -41,24 +41,27 @@ func NewCmdGenerate() *cobra.Command {
 	cmd.PersistentFlags().StringVarP(&raidType, "raidtype", "", "stripe",
 		"allowed RAID configuration such as, stripe, mirror, raid, raidz2")
+	cmd.PersistentFlags().StringVarP(&cap, "capacity", "", "10Gi",
+		"allowed RAID configuration such as, stripe, mirror, raid, raidz2")
```
Need to change helper text... minimum disk capacity to create pool.

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -181,48 +201,84 @@ func makePools(poolType string, nDevices int, bd map[string][]v1alpha1.BlockDevi
+			return nil, fmt.Errorf("number of devices must be a multiple of %d", min)
+		}
+		if min > nDevices {
+			return nil, fmt.Errorf("insufficient blockdevices expected for %s", poolType)
```
nit:
```suggestion
			return nil, fmt.Errorf("insufficient blockdevices expected %d no.of devices for %s", poolType)
```

vharsh replied:
this is thrown when the number-of-devices passed to the flag is lower than the minimum for a RaidGroup, I think I should handle this in the `cmd/generate/generate.go` itself.

mittachaitu commented on pkg/generate/sort.go:
```diff
@@ -0,0 +1,92 @@
+			prev.next = ahead.next
+			if count == 0 {
+				break
+			}
```
We are altering the count value and using it again at [end](https://github.com/openebs/openebsctl/pull/136/files#diff-43ff94923441f7c6da7fdb93c36630679fa70ce1a8680253660ce7e88e066eebR88) of func, Can we do in another way like below(Also fact that once it enters into this loop, for sure we had count no.of blockdevices)?
```go
prev.next = ahead.next
if len(results) == count {
   break
}
```

mittachaitu commented on pkg/generate/sort.go:
```diff
@@ -0,0 +1,92 @@
+			if count == 0 {
+				break
+			}
+		} else {
+			prev = curr
+			results = []v1alpha1.BlockDevice{}
+		}
+		curr = curr.next
+		ahead = ahead.next
```
nit: We can remove else part and merge with normal flow
```go
 prev = curr
 curr = curr.next
 ahead = ahead.next
```

vharsh replied:
Yeah, this makes sense considering I won't ever have a lesser number of BDs in the results slice, I'll either have all or none as the array is sorted.

vharsh replied:
I'll make the change and hope tests catch any potential change, on the cover this change makes super sense and shouldn't fail the tests.

mittachaitu commented on pkg/generate/sort.go:
```diff
@@ -0,0 +1,92 @@
+		return head, nil, fmt.Errorf("no blockdevices of equal sizes to select in a RaidGroup")
+	} else if len(results) < count {
+		return head, nil, fmt.Errorf("wanted %d blockdevices, have %d to pick", count, len(results))
+	}
```
nit: We can combine both something like below
```go
if len(results) != count {
   return head, nil, fmt.Errorf("wanted %d blockdevices, have %d to pick", count, len(results))
}
```

mittachaitu commented on pkg/generate/sort_test.go:
```diff
@@ -0,0 +1,160 @@
+			args{list: v1alpha1.BlockDeviceList{Items: []v1alpha1.BlockDevice{goodBD1N1, goodBD1N2, goodBD1N3, goodBD2N1, goodBD3N1}}},
+			&DeviceList{goodBD1N1, &DeviceList{goodBD1N2, &DeviceList{goodBD1N3,
+				&DeviceList{goodBD2N1, &DeviceList{goodBD3N1, nil}}}}},
+		},
```
Optional: Can we add one test with a different BD size I saw that all the BD's have the same storage.capacity value

vharsh replied:
Yeah, I've written a helper method which takes a slice of ints which are basically the gigabytes(powers of 10) sizes of the BDs objects so that I don't need to write them all over again.

mittachaitu commented on pkg/generate/sort.go:
```diff
@@ -0,0 +1,92 @@
+
+// Select returns count number of Blockdevices from the DeviceList LinkedList
+func (head *DeviceList) Select(size resource.Quantity, count int) (*DeviceList, []v1alpha1.BlockDevice, error) {
+	if size.Cmp(resource.MustParse("0")) == 0 {
+		return nil, nil, fmt.Errorf("size is zero")
+	}
```
Q: Is this check necessary? by considering the flow at caller of this func

vharsh replied:
Not necessary actually this must not even be tested, this is sort of a way to get a certain number of blockdevices of the same size without minimumSize restrictions(note: they'll still be >64MiB though, that's taken care of).

mittachaitu commented on pkg/generate/sort_test.go:
```diff
@@ -0,0 +1,160 @@
+			[]v1alpha1.BlockDevice{bdGen(2, 4), bdGen(3, 4), bdGen(4, 4)}, false},
+		{"six node LinkedList, three BD required of 1G", args{bdLinkedList(6, []int{5, 10, 15, 20, 25, 30}), resource.MustParse("1G"), 3},
+			nil, true},
+		{"six node LinkedList, three BD required of 1G", args{bdLinkedList(6, []int{1, 1, 10, 20, 25, 30}), resource.MustParse("1G"), 2},
```
```suggestion
		{"six node LinkedList, two BDs required of 1G", args{bdLinkedList(6, []int{1, 1, 10, 20, 25, 30}), resource.MustParse("1G"), 2},
```

mittachaitu commented on pkg/generate/sort_test.go:
```diff
@@ -0,0 +1,160 @@
+			nil, true},
+		{"six node LinkedList, three BD required of 1G", args{bdLinkedList(6, []int{1, 1, 10, 20, 25, 30}), resource.MustParse("1G"), 2},
+			[]v1alpha1.BlockDevice{bdGen(1, 1), bdGen(2, 1)}, false},
+		{"six node LinkedList, three BD required of 6G", args{bdLinkedList(6, []int{5, 10, 10, 20, 25, 30}), resource.MustParse("1G"), 2},
```
```suggestion
		{"six node LinkedList, two BD required of 6G", args{bdLinkedList(6, []int{5, 10, 10, 20, 25, 30}), resource.MustParse("1G"), 2},
```

mittachaitu commented on pkg/generate/sort_test.go:
```diff
@@ -0,0 +1,160 @@
+		{"six node LinkedList, three BD required of 1G", args{bdLinkedList(6, []int{1, 1, 10, 20, 25, 30}), resource.MustParse("1G"), 2},
+			[]v1alpha1.BlockDevice{bdGen(1, 1), bdGen(2, 1)}, false},
+		{"six node LinkedList, three BD required of 6G", args{bdLinkedList(6, []int{5, 10, 10, 20, 25, 30}), resource.MustParse("1G"), 2},
+			[]v1alpha1.BlockDevice{bdGen(2, 10), bdGen(3, 10)}, false},
```
Can we add a test case to pass invalid arguments to select function? i.e passing sizes in non-sorted order?

vharsh replied:
it'll fail with an error, good suggestion, it makes sense to also have a sane test case which is actually a negative one

vharsh replied:
this would fail, but it'd be a good candidate for testing, adding it and checking for a `true` error

vharsh replied:
added `six node LinkedList with unsorted BD sizes, two BD required of 1G`, this request fails even when the BDs are present because they don't come under the same window of size=2 together

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -181,48 +201,84 @@ func makePools(poolType string, nDevices int, bd map[string][]v1alpha1.BlockDevi
+			devices := Generate(v1alpha1.BlockDeviceList{Items: bds})
+			for d := 0; d < nDevices/min; d++ {
+				var raids []cstorv1.CStorPoolInstanceBlockDevice
+				tmp, thisRaidGroup, err := devices.Select(minsize, min)
+				// re-assign the head node of the linked-list for next iteration
+				//
+				devices = tmp
```
nit:
```suggestion
				devices, thisRaidGroup, err := devices.Select(minsize, min)
				// re-assign the head node of the linked-list for next iteration
				//
				devices = tmp
```

vharsh replied:
This single line changeset will fail the tests, if I do it, it'll create(thanks to the walrus operator `:=`) a `devices` variable in the scope of the for loop and will overshadow the `devices` in the scope just above it in hierarchy and it'll go out of scope, unless I do a `devices = devices` which seems a bit off.at first to most people, that's why the name `tmp`, I won't mind renaming this `tmp` to `d` as per Golang's naming convention, the length of the variable name should be as long as it remains in scope(since this one would be in scope for a very short time).

mittachaitu replied:
Sorry, I didn't observe at `:=`... My bad :(

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -181,48 +201,84 @@ func makePools(poolType string, nDevices int, bd map[string][]v1alpha1.BlockDevi
-	}
-	return nil, fmt.Errorf("unknown pool-type")
+	default:
+		return nil, fmt.Errorf("unknown pool-type")
```
Let's also log obtained pool type, for sake of completeness

mittachaitu reviewed (commented):
Few comments... logic wise looks good

Commit:
Update error messages, crunch code-branches, etc

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Allow 0 bytes as minimum disk size in Disk Selection algo

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

mittachaitu commented on pkg/generate/sort.go:
```diff
@@ -0,0 +1,86 @@
+
+// DeviceList is a LinkedList of BlockDevices
+type DeviceList struct {
+	item v1alpha1.BlockDevice
```
nit: If we made it as pointer copying of things will be reduced to great extent as compared to struct copy

mittachaitu reviewed (approved):
LGTM

vharsh commented:
Makes sense, I’ll have to change how unit tests check them internally, but
this seems doable.

On Fri, 24 Dec 2021 at 1:17 PM, sai chaithanya ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> LGTM
> ------------------------------
>
> In pkg/generate/sort.go
> <https://github.com/openebs/openebsctl/pull/136#discussion_r774914482>:
>
> > @@ -0,0 +1,86 @@
> +package generate
> +
> +import (
> +	"fmt"
> +
> +	"github.com/openebs/api/v2/pkg/apis/openebs.io/v1alpha1"
> +	"k8s.io/apimachinery/pkg/api/resource"
> +)
> +
> +// DeviceList is a LinkedList of BlockDevices
> +type DeviceList struct {
> +	item v1alpha1.BlockDevice
>
> nit: If we made it as pointer copying of things will be reduced to great
> extent as compared to struct copy
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openebs/openebsctl/pull/136#pullrequestreview-839789164>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACG7P4XKZYHES6UYJTURNB3USQQQ5ANCNFSM5JEF3M7Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mittachaitu commented:
> Makes sense, I’ll have to change how unit tests check them internally, but this seems doable.
> […](#)
> On Fri, 24 Dec 2021 at 1:17 PM, sai chaithanya ***@***.***> wrote: ***@***.**** approved this pull request. LGTM ------------------------------ In pkg/generate/sort.go <[#136 (comment)](https://github.com/openebs/openebsctl/pull/136#discussion_r774914482)>: > @@ -0,0 +1,86 @@ +package generate + +import ( + "fmt" + + "github.com/openebs/api/v2/pkg/apis/openebs.io/v1alpha1" + "k8s.io/apimachinery/pkg/api/resource" +) + +// DeviceList is a LinkedList of BlockDevices +type DeviceList struct { + item v1alpha1.BlockDevice nit: If we made it as pointer copying of things will be reduced to great extent as compared to struct copy — Reply to this email directly, view it on GitHub <[#136 (review)](https://github.com/openebs/openebsctl/pull/136#pullrequestreview-839789164)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACG7P4XKZYHES6UYJTURNB3USQQQ5ANCNFSM5JEF3M7Q> . You are receiving this because you were mentioned.Message ID: ***@***.***>

reflect.DeepEqual should work

Commit:
Make the DeviceList contain a pointer to BD

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
silence errcheck in benchmark tests

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

vharsh commented:
> reflect.DeepEqual should work

Yes, it did, made some minor changes in the unit test data. It likely also improved performance numbers.

vharsh commented:
Before:
```
BenchmarkSelect
BenchmarkSelect/six_node_LinkedList,_two_BD_required_of_6G
BenchmarkSelect/six_node_LinkedList,_two_BD_required_of_6G-8         	  282495	      4133 ns/op
BenchmarkSelect/six_node_LinkedList_with_unsorted_BD_sizes,_two_BD_required_of_1G
BenchmarkSelect/six_node_LinkedList_with_unsorted_BD_sizes,_two_BD_required_of_1G-8         	  232467	      5830 ns/op
PASS
```


After:
```
cpu: AMD Ryzen 7 4700U with Radeon Graphics         
BenchmarkSelect/six_node_LinkedList,_two_BD_required_of_6G-8         	  414178	      3910 ns/op	     848 B/op	      15 allocs/op
BenchmarkSelect/six_node_LinkedList_with_unsorted_BD_sizes,_two_BD_required_of_1G-8         	  282440	      5336 ns/op	     944 B/op	      23 allocs/op
PASS
ok  	github.com/openebs/openebsctl/pkg/generate	3.279s
```

mittachaitu reviewed (approved):
LGTM
